storage/database/db_migration.go

```diff
@@ -0,0 +1,120 @@
+// Modifications Copyright 2020 The klaytn Authors
+// Copyright 2015 The go-ethereum Authors
+// This file is part of go-ethereum.
+//
+// go-ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// go-ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with go-ethereum. If not, see <http://www.gnu.org/licenses/>.
+
+package database
+
+import (
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	dbMigrationFetchNum = 500
+)
+
+// StartDBMigration migrates a DB to a different kind of DB.
+// (e.g. LevelDB -> BadgerDB, LevelDB -> DynamoDB)
+//
+// This feature uses Iterator. A src DB should have implementation of Iteratee to use this function.
+// Do not use db migration while a node is executing.
+func (dbm *databaseManager) StartDBMigration(dstdbm DBManager) error {
+	// settings for quit signal from os
+	sigQuit := make(chan os.Signal, 1)
+	signal.Notify(sigQuit,
+		syscall.SIGHUP,
+		syscall.SIGINT,
+		syscall.SIGTERM,
+		syscall.SIGQUIT)
+
+	// TODO enable for all dbs
+	srcDB := dbm.getDatabase(MiscDB) // first DB
+	dstDB := dstdbm.getDatabase(MiscDB)
+
+	// create src iterator and dst batch
+	srcIter := srcDB.NewIterator()
+	dstBatch := dstDB.NewBatch()
+
+	// vars for log
+	start := time.Now()
+	iterateNum, fetchedTotal, fetched := 0, 0, 1
+
+loop:
+	for fetched > 0 {
+		// fetch keys and values
+		var keys, vals [][]byte
+		iterStart := time.Now()
+		keys, vals, fetched = iterateDB(srcIter, dbMigrationFetchNum)
```

Comment:
Why does it accumulate `keys` and `vals`? It could put items directly iterating DB.

Reply:
`iter.Key()` and `iter.Value()` is only available until `iter.next()` is called. If the items are put directly, the values is changed when `iter.next()` is called. The values should be stored somewhere.